NDTensors/src/zeros/allocate.jl

```diff
@@ -0,0 +1,36 @@
+function allocate(T::Tensor, elt::Type=default_eltype())
+  if !is_unallocated_zeros(T)
+    return T
+  end
+  ## allocate the tensor if is_unallocated_zeros
+  store = allocate(storage(T), elt)
+  return Tensor(NDTensors.AllowAlias(), store, inds(T))
+end
+
+function allocate(storage::TensorStorage, elt::Type=default_eltype())
+  if !is_unallocated_zeros(storage)
+    return storage
+  end
+  alloc = allocate(data(storage), elt)
+
+  #d = adapt(storage, typeof(alloc))(alloc)
+  return set_datatype(typeof(storage), typeof(alloc))(alloc)
+end
+
+function allocate(storage::BlockSparse, elt::Type=default_eltype())
+  if !is_unallocated_zeros(storage)
+    return storage
+  end
+  alloc = allocate(data(storage), elt)
+
+  #d = adapt(storage, typeof(alloc))(alloc)
```

Comment:
Seems like we could use:
```julia
function allocate(T::Tensor)
  return adapt(alloctype(data(T)), T)
end

function allocate(T::TensorStorage)
  return adapt(alloctype(data(T)), T)
end
```
If we define a generic fallback `alloctype(arraytype::Type{<:AbstractArray}) = arraytype`, which I think is reasonable to do since we assume a generic array type is allocated, then this should "just work" whether or not `T` has allocated data.

Reply:
This does work, though I think it would be good to verify the performance of tensors that are already allocated. Also I was able to get adapt to work to set the eltype after allocating UnallocatedZeros with UnspecifiedZero type.

Reply:
Great. Agreed we would want to check it doesn't hurt performance of allocated tensors. Ideally it is just a no-op but that assumes it gets optimized away.